File: backend/src/main/java/com/example/realestate/model/ProjectMilestone.java
package com.example.realestate.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;

import java.time.LocalDate;

@Entity
@Table(name = "project_milestones")
public class ProjectMilestone {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "project_id", nullable = false)
    private Long projectId;

    @Column(nullable = false)
    private String milestone;

    @Column(columnDefinition = "text")
    private String description;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    private LocalDate date;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MilestoneStatus status; // e.g., PENDING, COMPLETED

    public ProjectMilestone() {
    }

    public ProjectMilestone(Long projectId, String milestone, String description, LocalDate date, MilestoneStatus status) {
        this.projectId = projectId;
        this.milestone = milestone;
        this.description = description;
        this.date = date;
        this.status = status;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getProjectId() {
        return projectId;
    }

    public void setProjectId(Long projectId) {
        this.projectId = projectId;
    }

    public String getMilestone() {
        return milestone;
    }

    public void setMilestone(String milestone) {
        this.milestone = milestone;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public MilestoneStatus getStatus() {
        return status;
    }

    public void setStatus(MilestoneStatus status) {
        this.status = status;
    }
}

File: lib/models/project_milestone.dart
class ProjectMilestone {
  final int? id;
  final int projectId;
  final String milestone;
  final String? description;
  final String? date; // yyyy-MM-dd
  final String status; // PENDING, COMPLETED, CANCELLED

  ProjectMilestone({this.id, required this.projectId, required this.milestone, this.description, this.date, required this.status});

  factory ProjectMilestone.fromJson(Map<String, dynamic> json) => ProjectMilestone(
        id: json['id'] as int?,
        projectId: json['projectId'] ?? json['project_id'],
        milestone: json['milestone'] as String,
        description: json['description'] as String?,
        date: json['date'] as String?,
        status: json['status']?.toString() ?? 'PENDING',
      );

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'projectId': projectId,
        'milestone': milestone,
        if (description != null) 'description': description,
        if (date != null) 'date': date,
        'status': status,
      };
}
